packaging support using cpack

beadon opened the pull request:
- 06-29-2023 14:50 PST beadon  - script created for building and installation
- installation script updates
- builds packages now with cpack, name of packages not yet reliable
- 07-09-2023 21:55 beadon - updated readme to include amd64 packing command - updated cmake configuration file to support cpack DEB generator

beadon committed:
06-29-2023 14:50 PST beadon
 - script created for building and installation

beadon committed:
installation script updates

beadon committed:
builds packages now with cpack, name of packages not yet reliable

beadon committed:
07-09-2023 21:55 beadon
- updated readme to include amd64 packing command
- updated cmake configuration file to support cpack DEB generator

beadon committed:
Merge pull request #1 from hoehermann/whatsmeow

trying a sample PR from upstream ...

hoehermann commented:
Thank you. I want to look into this within a fortnight.

beadon commented:
note the versioning addition here to the cmake config.  This makes packaging and dependency checking easier.

hoehermann commented on CMakeLists.txt:
```diff
@@ -1,6 +1,16 @@
 ﻿cmake_minimum_required(VERSION 3.18) # needed on win32
 
-project("purple-gowhatsapp" LANGUAGES C)
+set(PROJECT_NAME "purple-gowhatsapp" C)
```
What is the purpose of defining the project name this way? It looks like the the "LANGUAGES" option is dropped and building fails on systems without a C++ compiler.

beadon replied:
An oversight if this breaks or removes functionality, I was simply following the skype4pidging packaging.  I didn't intend to modify the build!

I suspect we can keep the original code -- note the later reference to PROJECT_NAME will be removed, and since it's never used again -- it only appears after being set and then used in CMAKE_PROJECT_NAME for packages and file names later.

Thoughts ?

hoehermann replied:
I reverted to the simple `project(…)` command. It looks like it is working on all of my machines (Debian, arch and win32).

hoehermann commented on CMakeLists.txt:
```diff
@@ -1,6 +1,16 @@
 
-project("purple-gowhatsapp" LANGUAGES C)
+set(PROJECT_NAME "purple-gowhatsapp" C)
+set(VERSION_MAJOR "1")
+set(VERSION_MINOR "12")
+set(VERSION_PATCH "0")
+set(VERSION "${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}")
```
The version is already available in [VERSION](https://raw.githubusercontent.com/hoehermann/purple-gowhatsapp/whatsmeow/VERSION). I would rather like to read it from there so I do not need to maintain the version string in more than one place. Some folks check against the file rather than github tags.

Please note that the version of the build is actually more complicated. Ideally, it should include the whatsmeow version, as outlined in `version.sh`. whatsmeow is statically linked "embedded" into the build. Behavior and compatibility may differ, so the version string should reflect it. I do not even know if Debian packages offer a field for that.

beadon replied:
Yes, the version field is used to determine compatibility and upgrades.  To support upgrades from one version to the next my impression is that this is required.  Right now I am forming the VERSION string based on the major minor and patch versions because I envision a time when cmake-controlled build or package instructions require something version specific - ( perhaps to absorb a dependency of some kind, perhaps to drop features which require compile flags, etc..).  I have not fully explored the upgrade/downgrade path with dpkg -- you see I just merge everything into VERSION when I pump out the package name later.  Seemed like a light lfit to track versioning in one place close to the build tool ?

Since the major, minor and path versions are called out here, it makes these explicit -- my mind wanders to supporrting this with some kind of automatic git-commit versioning to make version tracking easier, but I am stuck with the excellet SemVer https://semver.org/ , which aways will require some light human touch to manage the complexity and breaking changes.

I see the challenge of the version field and showing / wrapping whatsmeow.  I would argue that since the version we are defining here is a THING, we can version it based on the tested dependencies and basically ignore the whatsmeow version -- because this is simply meta-data about what version was used with this version of purple-gowhatsapp.  This idea being - a version should always be atomic, if it happens to work with different versions of whatsmeow, then that's OK, but the version of whatsmeow is simply an integration test and a version bump of this code away from compability.

Are there places that https://raw.githubusercontent.com/hoehermann/purple-gowhatsapp/whatsmeow/VERSION is used other than ?

beadon replied:
OH -- correction, I already thought of this , check out : set(CPACK_PACKAGE_VERSION_MAJOR "${VERSION_MAJOR}")  which is injecting it in the dependency management controls of the packaging system(s).  I would strongly recommend we keep the broken-out-versions to keep dependency tracking in place..

hoehermann replied:
I try to use semantic versioning, although my minor version changes sometimes are not backwards-compatible.

`VERSION` is read by CMake, turned into a define for use in the plug-in. Additionally, it is relevant for PKGBUILD for arch. And now also in the cpack commands for Debian. :)

beadon replied:
Got it, so it is used elsewhere, then we will need to keep it.  Since it is defined outside of the dependencies within cmake, then the version in VERSION needs to be kept controlled.

I spotted a commit recently where there was a version downgrade in arch(?) packaging to align to the 1.12 version string.   It has me wondering if cmake can handle the arch packaging too so that it's passed from a central location and not changed in multiple places.  I'll see if I can find that recent commit ..

hoehermann replied:
You are probably referring to [this commit](https://aur.archlinux.org/cgit/aur.git/commit/?h=purple-gowhatsapp-git&id=e8f61c5286359a3f75975c0e323595c066ec756b) to the AUR. It was meant to be a local commit to be squashed later. Unfortunately, I pushed it early and had to undo the changes in a backwards compatible manner. As far as I know, AUR is the way to provide packages to Arch. I could remove the PKGCONF file from this repository, then it would be only on one place.

beadon replied:
I have not found a way to package for ARCH yet, but it will simplify things a great deal if all packaging runs through cmake/cpack ...  doing some more research....

And, yes, confirmation that this was the commit I saw bumping things to a 1.13 version.  if there is a bad package or version conflict due to packaging I recommend bumping the patch version.  We are technically managing 2 versions 1 version for the code, and another version for the packaging on the code.  Either could break, but hopefully the packaging is binary compatible so only requires a PATCH version bump.

This decision has downstream impact which means bumping the package version by a patch version will re-create packages on other platforms which have no changes -- in your example the ARCH package had a problem and needed to be rebuilt, and with this solution the DEB package would *also* get a patch version bump in addition to the ARCH package, but I am OK with the version control managed like this because over time this brings packaged versions together rather than having each platform's version diverge independently (difficult to maintain!).

Comments, questions, concerns ?

hoehermann replied:
I do not understand. While developing this plug-in, I usually have bursts of productivity that lead to new features being implemented. I push them to the main branch (whatsmeow). Then I wait for the early adopters who build from source to give me some feed-back. After things have been calm for two weeks, I bump the minor version update the change-log and create a tag. That is "the release" of the plug-in.

It is the same for all targeted platforms (Ubuntu 22.04, win32, Arch). There are no other versions or branches that I know of. No platform-specific versions.

There is, however, the version of the included whatsmeow release. That is managed by the go module system. I configured it to always use the latest version automatically. Since I do not want to update it manually, I want to keep it that way.

beadon replied:
Oh, we are talking about different things.

I saw a version applied to the purple-gowhatsapp(?) package for ARCH that was 1.13.x which appeared to be manually edited.  I was concerned that this was different than the cpack method for DEB packages which now rely on the VERSION file and are version 1.12.x.

How can we make the ARCH and DEB packaging versions consistent ?

hoehermann replied:
Yes, I edited the version string in the PKGBUILD file for Arch manually while preparing a future release. After everything was working fine, I accidentally pushed those changes. That was a mistake. I am sorry. In normal operation, the exact version string is automatically generated from the contents of VERSION. Then Debian and Arch is consistent.

hoehermann commented on CMakeLists.txt:
```diff
@@ -1,6 +1,16 @@
+
+# For now, we won't be doing any cross-compiling, so specifying amd64 here will cover most cases.
+# If you have a different processor and need to cross-compile, say, ARM, then this won't build the package you need until it's modified.
+set(ARCHITECTURE "amd64")
```
What is the purpose of setting an this variable? I cannot find any documentation about it. Does it affect CMake's choice of platform? If so, builds on win32 would fail.

beadon replied:
This is just a variable i use here to package later.  The build will not break on windows or other architectures.

The solution I propose here is not very scalable.

Perhaps I re-work it to detect the OS that we need to support and package for?  Unfortunately this may bleed into the build .  It appears that CMAKE requires the details of pieces like CMAKE_SYSTEM_PROCESSOR to truly change the architecture.  More research required so I can support build/package behavior without this variable ..

beadon replied:
going to resolve this, I can see that we are in agreement for managing th is next to the packaging location.  no longer an issue.

hoehermann commented on CMakeLists.txt:
```diff
@@ -37,3 +47,23 @@ add_subdirectory(src/go)
+set(CPACK_DEBIAN_PACKAGE_ARCHITECTURE "${ARCHITECTURE}")
+set(CPACK_SOURCE_PACKAGE_FILE_NAME "${CMAKE_PROJECT_NAME}_${VERSION}_${CPACK_DEBIAN_PACKAGE_ARCHITECTURE}")
+set(CPACK_PACKAGE_FILE_NAME "${CMAKE_PROJECT_NAME}_${VERSION}_${CPACK_DEBIAN_PACKAGE_ARCHITECTURE}")
+set(CPACK_DEBIAN_PACKAGE_DEPENDS "libpurple0 (>= 2.10.0), libglib2.0-0 (>= 2.24), libjson-glib-1.0-0 (>= 0.8.0)")
```
`libjson-glib-1.0-0` is not a dependency of this project. 

However, `gdk-pixbuf-2.0` and `opusfile` are optional dependencies which may or may not be present depending on what was available at time of cmake configuraion. This would need to be reflected in the list of dependencies.

Also, I am not sure about the versions. There are a couple of version switches in the code:

* GLIB_CHECK_VERSION in bridge.c and handle_attachment.c
* PURPLE_VERSION_CHECK in init.c handle_attachment.c
* maybe others somewhere else

These, too, affect (minimum) dependency versions during build time.

beadon replied:
Ok, then that means we are getting more complex in the building and packaging.  I would need to detect the build instructions and package different based on these dependencies. This could be done.

However, to make this a easier, I propose making `gdk-pixbuf-2.0` and `opusfile` requirements of the project.

Otherwise, if this kind of option is required, this balloons the number of packages from 1 to 4.  I struggle to see the value of publishing 4 different versions per platform ( without, with pixbuf, with opusfile, with pixbuf + opusfile ).  Then, arm , another 4 packages, arch, another 4  .. windows, another 4 ... then multiple this by the versions , and test every one on all these platforms.

If we think about this critically are there *real* active use cases where there is a need to leave these (optional) dependencies out ?

beadon replied:
I can easily remove the dependencies that are not required "libjson-glib-1.0-0 ".

hoehermann replied:
For the versions, I now use an approach that relies on the version information supplied by pkg-config.

The availability of optional libraries is checked, the dependencies are expanded conditionally.

beadon replied:
Ok, let's see how this goes.

I am concerned about the testing load this persists.  I would like to simplify packaging / building to a single set of dependencies long term.

hoehermann commented on CMakeLists.txt:
```diff
@@ -37,3 +47,23 @@ add_subdirectory(src/go)
 install(DIRECTORY "pixmaps" DESTINATION "${PURPLE_DATA_DIR}" FILES_MATCHING PATTERN "*.png")
+
+# package settings
+set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "Whatsapp protocol plug-in for libpurple with pidgin")
+set(CPACK_PACKAGE_VENDOR "Bryant Eadon")
+set(CPACK_PACKAGE_DESCRIPTION "whatsmeow is a go-based whatsapp protocol plug-in for libpurple based on the whatsapp business apis.")
+set(CPACK_PACKAGE_CONTACT "bryant@balancebitsconsulting.com")
+set(CPACK_PACKAGE_VERSION_MAJOR "${VERSION_MAJOR}")
+set(CPACK_PACKAGE_VERSION_MINOR "${VERSION_MINOR}")
+set(CPACK_PACKAGE_VERSION_PATCH "${VERSION_PATCH}")
+set(CPACK_DEBIAN_PACKAGE_SHLIBDEPS ON)
+set(CPACK_DEBIAN_PACKAGE_ARCHITECTURE "${ARCHITECTURE}")
+set(CPACK_SOURCE_PACKAGE_FILE_NAME "${CMAKE_PROJECT_NAME}_${VERSION}_${CPACK_DEBIAN_PACKAGE_ARCHITECTURE}")
+set(CPACK_PACKAGE_FILE_NAME "${CMAKE_PROJECT_NAME}_${VERSION}_${CPACK_DEBIAN_PACKAGE_ARCHITECTURE}")
```
This is pretty lengthy and will get longer as we consider more use-cases. I would like to have it in a separate file. I propose an outline like [this](https://github.com/hoehermann/purple-gowhatsapp/tree/cpack). Does that look okay to you?

beadon replied:
I can;t open the example you have unfortunately -- I have seen no example of people doing this in separate files.

The packaging files are very tightly coupled with build instructions, versions and more refer to how the binaries are built.  If it feels like clutter, perhaps there is a way to roll it up in an editor like VSCode ? or vim folds ?

All guides I have read so far are recommending including CPACK at the end of the CMakeLists.txt.

hoehermann replied:
I am one of those weird guys who code with a dumb text editor (syntax highlighting only). I moved the cpack commands into a separate file which is included by the main `CMakeLists.txt`. Seems to work so far.

beadon replied:
yes, that's fine.  

As for editors and style choices, I find a longer make file with everything in it is a reminder of the variables that you have at your fingertips.  Either other targets already built, or style choices you should be matching when writing new/different targets.

Anyway, this include method should work fine.

hoehermann commented on CMakeLists.txt:
```diff
@@ -37,3 +47,23 @@ add_subdirectory(src/go)
+set(CPACK_DEBIAN_PACKAGE_DEPENDS "libpurple0 (>= 2.10.0), libglib2.0-0 (>= 2.24), libjson-glib-1.0-0 (>= 0.8.0)")
+set(CPACK_DEBIAN_PACKAGE_MAINTAINER "Bryant Eadon") #required
+
+set(PACK "DEB" CACHE STRING "Generate a Package")
+set(CPACK_GENERATOR ${PACK})
```
Is there a reason we need `${PACK}` and not write `${CPACK_GENERATOR}` directly?

beadon replied:
Not specifically -- It was the code I was following from the skype4pidgin plugin.  I am sure there will be a reason to break it up into 2 lines at some later point as we build other packages, but for now I agree it is overly verbose.

I have not attempted to pack by combining the lines, but I imagine it would work just fine.  Would you like to see the 2 lines collapsed ?

hoehermann replied:
Yes, I prefer collapsed lines (done).

beadon replied:
k, resolved.

hoehermann commented:
I think I understood most of the changes. I rebased onto the main branch and squashed commits to my liking. Thank you for the contribution.

beadon commented:
> I think I understood most of the changes. I rebased onto the main branch and squashed commits to my liking. Thank you for the contribution.

I appreciate the acceptance!

I raised some points to agree for - versioning, and dependencies.  Can you take a quick look at these to see if the proposal make sense ?  I don't want to break you/repo workflow with a bad suggestion/implementation.
